geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/DescriptiveStats.scala

```diff
@@ -0,0 +1,358 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.utils.stats
+
+import org.ejml.data.DenseMatrix64F
+import org.ejml.ops.CommonOps
+import org.ejml.simple.SimpleMatrix
+import org.locationtech.geomesa.utils.stats.SimpleMatrixUtils._
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.collection.JavaConverters._
+import scala.collection.immutable.ListMap
+import scala.Array._
+
+class DescriptiveStats(val attributes: Seq[Int]) extends Stat with Serializable {
+
+  override type S = DescriptiveStats
+
+  private[stats] val size = attributes.size
+  private[stats] val size_squared = size * size
+
+  private[stats] var _count: Long = _
+  private[stats] val _min: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _max: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _sum: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _mean: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _m2n: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _m3n: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _m4n: SimpleMatrix = new SimpleMatrix(size, 1)
+  private[stats] val _c2: SimpleMatrix = new SimpleMatrix(size, size)
+
+  clear()
+
+  override def clear(): Unit = {
+    _count = 0
+    _min.set(java.lang.Double.MAX_VALUE)
+    _max.set(0d - java.lang.Double.MAX_VALUE)
+    _sum.set(0d)
+    _mean.set(1d)
+    _m2n.set(0d)
+    _m3n.set(0d)
+    _m4n.set(0d)
+    _c2.set(0d)
+  }
+
+  def copyFrom(that: DescriptiveStats): Unit = {
+    _count = that._count
+    _min.set(that._min)
+    _max.set(that._max)
+    _sum.set(that._sum)
+    _mean.set(that._mean)
+    _m2n.set(that._m2n)
+    _m3n.set(that._m3n)
+    _m4n.set(that._m4n)
+    _c2.set(that._c2)
+  }
+
+  def count: Long = _count
+
+  def min: Array[Double] = (if (isEmpty) _max else _min).getMatrix.data.clone()
+  def max: Array[Double] = (if (isEmpty) _min else _max).getMatrix.data.clone()
```

Comment:
it does copy, but only when method is called with should typically be once per aggregate calculation.  i really don't want to leak mutable constructs...